src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1580,6 +1604,17 @@ public static unsafe SocketError GetSockOpt(SafeSocketHandle handle, SocketOptio
             int optLen = sizeof(int);
             Interop.Error err = Interop.Sys.GetSockOpt(handle, optionLevel, optionName, (byte*)&value, &optLen);
 
+#if SYSTEM_NET_SOCKETS_APPLE_PLATFROM
+            // macOS fails to even query it if socket is not actively listening.
+            // To provide consistent platform experience we will track if
+            // it was ret and we will use it later as needed.
+            if (optionLevel == SocketOptionLevel.Tcp && optionName == SocketOptionName.FastOpen && err != Interop.Error.SUCCESS)
+            {
+                value = handle.TfoEnabled ? 1 : 0;
+                err = Interop.Error.SUCCESS;
+            }
```

Comment:
this is there on purpose. We try the real call first and it it wrks great. We return the real thing. 
If not, we fallback to this to avoid user needing to catch the particular error.  With this it works for server listener as normal, and it can light tp on client if Apple ever improves the support.